File: matlab/+capstone/+io/Protocol.m
classdef Protocol
    % Protocol constants (ESP32 CMD list aligned; MATLAB legacy aliases kept)
    methods (Static)
        function p = constants()
            p = struct();

            % ===================== MODES / GLOBAL (SYSTEM) =====================
            % Khớp với enum Cmd trong protocol.hpp:
            %   CMD_GLOBAL_OPERATION      = 0x01
            %   CMD_GLOBAL_IDLE           = 0x02
            %   CMD_GLOBAL_PING_MODE      = 0x03
            %   CMD_GLOBAL_EMRGENCY_STOP  = 0x04
            p.CMD_MODE_OP                = uint8(hex2dec('01')); % CMD_GLOBAL_OPERATION
            p.CMD_MODE_IDLE              = uint8(hex2dec('02')); % CMD_GLOBAL_IDLE
            p.CMD_GLOBAL_PING_MODE       = uint8(hex2dec('03')); % CMD_GLOBAL_PING_MODE
            p.CMD_GLOBAL_EMERGENCY_STOP  = uint8(hex2dec('04')); % CMD_GLOBAL_EMRGENCY_STOP

            % ACK byte từ ESP32 sau SYSTEM + IDLE
            p.ACK                        = uint8(hex2dec('20')); % 0x20

            % ===================== OPERATION (ESP32) =====================
            % Khớp với:
            %   CMD_OP_PWM_FWD        = 0x10
            %   CMD_OP_PWM_BWD        = 0x11
            %   CMD_OP_SPD_FWD        = 0x12
            %   CMD_OP_SPD_BWD        = 0x13
            %   CMD_OP_BRAKE          = 0x14
            %   CMD_OP_LEGACY_CTRL_5B = 0x15
            p.CMD_OP_PWM_FWD        = uint8(hex2dec('10')); % payload: u16 (11-bit duty) LE
            p.CMD_OP_PWM_BWD        = uint8(hex2dec('11')); % payload: u16 (11-bit duty) LE
            p.CMD_OP_SPD_FWD        = uint8(hex2dec('12')); % payload: float32 (speed) LE
            p.CMD_OP_SPD_BWD        = uint8(hex2dec('13')); % payload: float32 (speed) LE
            p.CMD_OP_BRAKE          = uint8(hex2dec('14')); % payload: none
            p.CMD_OP_LEGACY_CTRL_5B      = uint8(hex2dec('15')); % speed_u16 + angle_u16 (LE)

            % ===================== IDLE SENSORS (ESP32) =====================
            % Khớp với:
            %   CMD_IDLE_SENSOR_LINE_READ   = 0x20
            %   CMD_IDLE_SENSOR_ULTRA_READ  = 0x21
            %   CMD_IDLE_SENSOR_ULTRA_KICK  = 0x22
            %   CMD_IDLE_SENSOR_MPU_READ    = 0x23
            %   CMD_IDLE_ENCODER_ENABLE     = 0x24
            %   CMD_IDLE_ENCODER_DISABLE    = 0x25
            p.CMD_IDLE_SENSOR_LINE_READ   = uint8(hex2dec('20'));
            p.CMD_IDLE_SENSOR_ULTRA_READ  = uint8(hex2dec('21'));
            p.CMD_IDLE_SENSOR_ULTRA_KICK  = uint8(hex2dec('22'));
            p.CMD_IDLE_SENSOR_MPU_READ    = uint8(hex2dec('23'));
            p.CMD_IDLE_ENCODER_ENABLE     = uint8(hex2dec('24'));
            p.CMD_IDLE_ENCODER_DISABLE    = uint8(hex2dec('25'));

            % ===================== IDLE ACTUATORS (ESP32) =====================
            % Khớp với:
            %   CMD_IDLE_ACTUATOR_MOTOR_ENABLE   = 0x30
            %   CMD_IDLE_ACTUATOR_MOTOR_DISABLE  = 0x31
            %   CMD_IDLE_ACTUATOR_MOTOR_PWM_FWD  = 0x32
            %   CMD_IDLE_ACTUATOR_MOTOR_PWM_BWD  = 0x33
            %   CMD_IDLE_ACTUATOR_MOTOR_SPD_FWD  = 0x34
            %   CMD_IDLE_ACTUATOR_MOTOR_SPD_BWD  = 0x35
            %   CMD_IDLE_ACTUATOR_MOTOR_STOP     = 0x36
            %   CMD_IDLE_ACTUATOR_SERVO_ENABLE        = 0x40
            %   CMD_IDLE_ACTUATOR_SERVO_DISABLE       = 0x41
            %   CMD_IDLE_ACTUATOR_SERVO_WRITE         = 0x42
            %   CMD_IDLE_ACTUATOR_SERVO_READ          = 0x43
            %   CMD_IDLE_ACTUATOR_SERVO_WRITE_CENTER  = 0x44
            p.CMD_IDLE_ACTUATOR_MOTOR_ENABLE      = uint8(hex2dec('30'));
            p.CMD_IDLE_ACTUATOR_MOTOR_DISABLE     = uint8(hex2dec('31'));
            p.CMD_IDLE_ACTUATOR_MOTOR_PWM_FWD     = uint8(hex2dec('32')); % payload: u16 (LE)
            p.CMD_IDLE_ACTUATOR_MOTOR_PWM_BWD     = uint8(hex2dec('33')); % payload: u16 (LE)
            p.CMD_IDLE_ACTUATOR_MOTOR_SPD_FWD     = uint8(hex2dec('34')); % payload: u16 (LE)
            p.CMD_IDLE_ACTUATOR_MOTOR_SPD_BWD     = uint8(hex2dec('35')); % payload: u16 (LE)
            p.CMD_IDLE_ACTUATOR_MOTOR_STOP        = uint8(hex2dec('36'));

            p.CMD_IDLE_ACTUATOR_SERVO_ENABLE      = uint8(hex2dec('40'));
            p.CMD_IDLE_ACTUATOR_SERVO_DISABLE     = uint8(hex2dec('41'));
            p.CMD_IDLE_ACTUATOR_SERVO_WRITE       = uint8(hex2dec('42')); % payload: angle_u16 (LE)
            p.CMD_IDLE_ACTUATOR_SERVO_READ        = uint8(hex2dec('43'));
            p.CMD_IDLE_ACTUATOR_SERVO_WRITE_CENTER= uint8(hex2dec('44'));

            % ===================== IDLE SET PARAMS (payload fixed 30) =====================
            % Khớp với:
            %   CMD_IDLE_SET_LINE_PARAMS   = 0x50
            %   CMD_IDLE_SET_MPU_PARAMS    = 0x51
            %   CMD_IDLE_SET_ULTRA_PARAMS  = 0x52
            %   CMD_IDLE_SET_MOTOR_PARAMS  = 0x53
            %   CMD_IDLE_SET_SERVO_PARAMS  = 0x54
            %   CMD_IDLE_SET_PID_PARAMS    = 0x55
            p.CMD_IDLE_SET_LINE_PARAMS   = uint8(hex2dec('50'));
            p.CMD_IDLE_SET_MPU_PARAMS    = uint8(hex2dec('51'));
            p.CMD_IDLE_SET_ULTRA_PARAMS  = uint8(hex2dec('52'));
            p.CMD_IDLE_SET_MOTOR_PARAMS  = uint8(hex2dec('53'));
            p.CMD_IDLE_SET_SERVO_PARAMS  = uint8(hex2dec('54'));
            p.CMD_IDLE_SET_PID_PARAMS    = uint8(hex2dec('55'));

            % ===================== LEGACY MATLAB NAMES (ALIASES) =====================
            % Giữ alias cũ để không phải sửa toàn bộ code MATLAB:
            p.CMD_READ_LINE   = p.CMD_IDLE_SENSOR_LINE_READ;
            p.CMD_READ_ULTRA  = p.CMD_IDLE_SENSOR_ULTRA_READ;
            p.CMD_READ_MPU    = p.CMD_IDLE_SENSOR_MPU_READ;

            p.CMD_DRIVE_FWD      = p.CMD_IDLE_ACTUATOR_MOTOR_SPD_FWD;
            p.CMD_DRIVE_BWD      = p.CMD_IDLE_ACTUATOR_MOTOR_SPD_BWD;
            p.CMD_DRIVE_STOP     = p.CMD_IDLE_ACTUATOR_MOTOR_STOP;
            p.CMD_STEER_DISABLE  = p.CMD_IDLE_ACTUATOR_SERVO_DISABLE;
            p.CMD_STEER_ANGLE    = p.CMD_IDLE_ACTUATOR_SERVO_WRITE;

            % ===================== FRAMES =====================
            p.SENSOR_FRAME_LEN = uint16(22); % OPERATION telemetry frame length
        end
    end
end

File: low_level/src/protocol/protocol.cpp
#include "protocol.hpp"
#include <string.h>

namespace Protocol {

Group scan(Cmd cmd)
{
  if (cmd == Cmd::CMD_GLOBAL_OPERATION ||
      cmd == Cmd::CMD_GLOBAL_IDLE ||
      cmd == Cmd::CMD_GLOBAL_PING_MODE ||
      cmd == Cmd::CMD_GLOBAL_EMRGENCY_STOP) {
    return Group::SYSTEM;
  }

  if (cmd == Cmd::CMD_OP_PWM_FWD ||
      cmd == Cmd::CMD_OP_PWM_BWD ||
      cmd == Cmd::CMD_OP_SPD_FWD ||
      cmd == Cmd::CMD_OP_SPD_BWD ||
      cmd == Cmd::CMD_OP_BRAKE   ||
      cmd == Cmd::CMD_OP_LEGACY_CTRL_5B) {
    return Group::OPERATION;
  }

  switch (cmd) {
    case Cmd::CMD_IDLE_SENSOR_LINE_READ:
    case Cmd::CMD_IDLE_SENSOR_ULTRA_READ:
    case Cmd::CMD_IDLE_SENSOR_ULTRA_KICK:
    case Cmd::CMD_IDLE_SENSOR_MPU_READ:
    case Cmd::CMD_IDLE_ENCODER_ENABLE:
    case Cmd::CMD_IDLE_ENCODER_DISABLE:
    case Cmd::CMD_IDLE_ACTUATOR_MOTOR_ENABLE:
    case Cmd::CMD_IDLE_ACTUATOR_MOTOR_DISABLE:
    case Cmd::CMD_IDLE_ACTUATOR_MOTOR_PWM_FWD:
    case Cmd::CMD_IDLE_ACTUATOR_MOTOR_PWM_BWD:
    case Cmd::CMD_IDLE_ACTUATOR_MOTOR_SPD_FWD:
    case Cmd::CMD_IDLE_ACTUATOR_MOTOR_SPD_BWD:
    case Cmd::CMD_IDLE_ACTUATOR_MOTOR_STOP:
    case Cmd::CMD_IDLE_ACTUATOR_SERVO_ENABLE:
    case Cmd::CMD_IDLE_ACTUATOR_SERVO_DISABLE:
    case Cmd::CMD_IDLE_ACTUATOR_SERVO_WRITE:
    case Cmd::CMD_IDLE_ACTUATOR_SERVO_READ:
    case Cmd::CMD_IDLE_ACTUATOR_SERVO_WRITE_CENTER:
    case Cmd::CMD_IDLE_SET_LINE_PARAMS:
    case Cmd::CMD_IDLE_SET_MPU_PARAMS:
    case Cmd::CMD_IDLE_SET_ULTRA_PARAMS:
    case Cmd::CMD_IDLE_SET_MOTOR_PARAMS:
    case Cmd::CMD_IDLE_SET_SERVO_PARAMS:
    case Cmd::CMD_IDLE_SET_PID_PARAMS:
      return Group::IDLE;
    default:
      return Group::UNKNOWN;
  }
}

Payload payloadOf(Cmd cmd)
{
  switch (cmd) {
    case Cmd::CMD_OP_PWM_FWD:
    case Cmd::CMD_OP_PWM_BWD:
    case Cmd::CMD_IDLE_ACTUATOR_SERVO_WRITE:
    case Cmd::CMD_IDLE_ACTUATOR_MOTOR_PWM_FWD:
    case Cmd::CMD_IDLE_ACTUATOR_MOTOR_PWM_BWD:
    case Cmd::CMD_IDLE_ACTUATOR_MOTOR_SPD_FWD:
    case Cmd::CMD_IDLE_ACTUATOR_MOTOR_SPD_BWD:
      return Payload::U16_LE;

    case Cmd::CMD_OP_SPD_FWD:
    case Cmd::CMD_OP_SPD_BWD:
        return Payload::F32_LE;

    case Cmd::CMD_IDLE_SET_LINE_PARAMS:
    case Cmd::CMD_IDLE_SET_MPU_PARAMS:
    case Cmd::CMD_IDLE_SET_ULTRA_PARAMS:
    case Cmd::CMD_IDLE_SET_MOTOR_PARAMS:
    case Cmd::CMD_IDLE_SET_SERVO_PARAMS:
    case Cmd::CMD_IDLE_SET_PID_PARAMS:
      return Payload::BYTES30;

    case Cmd::CMD_OP_LEGACY_CTRL_5B:
      return Payload::OP5B;

    default:
      return Payload::NONE;
  }
}

uint8_t payloadLen(Payload p)
{
    switch (p) {
      case Payload::NONE:    return 0;
      case Payload::U16_LE:  return 2;
      case Payload::BYTES30: return 30;
      case Payload::OP5B:    return 4;
      case Payload::F32_LE:  return 4;   // float32
      default:               return 0;
    }
}


bool tryRead(Network& net, RxFrame& out)
{
  uint8_t cmd_u8 = 0;
  if (!net.getUint8(cmd_u8)) {
    return false;
  }

  Cmd     cmd = (Cmd)cmd_u8;
  Payload pl  = payloadOf(cmd);

  out.cmd     = cmd;
  out.group   = scan(cmd);
  out.payload = pl;
  out.u16     = 0;
  out.speed_u16 = 0;
  out.angle_u16 = 0;
  memset(out.bytes30, 0, sizeof(out.bytes30));

  switch (pl) {
    case Payload::NONE:
      break;

    case Payload::U16_LE: {
      uint8_t buf[2];
      if (!net.getArrayUint8(buf, 2)) return false;
      out.u16 = get_u16_le(buf);
      break;
    }

    case Payload::BYTES30: {
      if (!net.getArrayUint8(out.bytes30, 30)) return false;
      break;
    }

    case Payload::OP5B: {
      uint8_t buf[4];
      if (!net.getArrayUint8(buf, 4)) return false;
      out.speed_u16 = get_u16_le(&buf[0]);
      out.angle_u16 = get_u16_le(&buf[2]);
      break;
    }

    case Payload::F32_LE: {
      uint8_t buf[4];
      if (!net.getArrayUint8(buf, 4)) return false;
      float v;
      memcpy(&v, buf, 4);   // ESP32 little-endian nên copy thẳng
      out.f32 = v;
      break;
    }

    default:
      break;
  }

  return true;
}

} // namespace Protocol
